### ISSUE TITLE
Are there any plans to support context parallel?

### ISSUE THREAD

**Issue by dz1iang** [pytorch/torchtune#2141]
Long-text scenarios are quite common, and it would be of great help if they could be supported.

**Comment by joecummings** [comment]
Take a look at @felipemello1's awesome RFC looking at how we would plan to support even longer context models: #1244. 

We started taking a look at it, but then de-prioritized the work in favor of onboarding new modalities b/c we found that with our memory optimizations we could easily get to 64K context length. 

What use case are you trying to work on? We could revisit our prioritization.

**Comment by dz1iang** [comment]
> Take a look at @felipemello1's awesome RFC looking at how we would plan to support even longer context models: #1244.
> 
> We started taking a look at it, but then de-prioritized the work in favor of onboarding new modalities b/c we found that with our memory optimizations we could easily get to 64K context length.
> 
> What use case are you trying to work on? We could revisit our prioritization.

In my opinion, everyone is chasing after O1. In this process, the training on long texts in SFT (Supervised Fine-Tuning) and RL (Reinforcement Learning) is necessary.

**Comment by xs1997zju** [comment]
> Take a look at @felipemello1's awesome RFC looking at how we would plan to support even longer context models: #1244.
> 
> We started taking a look at it, but then de-prioritized the work in favor of onboarding new modalities b/c we found that with our memory optimizations we could easily get to 64K context length.
> 
> What use case are you trying to work on? We could revisit our prioritization.

@joecummings Does it support  128k of 70b size model full training? I mean, not lora, but full parameter

**Comment by xs1997zju** [comment]
> > Take a look at @felipemello1's awesome RFC looking at how we would plan to support even longer context models: #1244.
> > We started taking a look at it, but then de-prioritized the work in favor of onboarding new modalities b/c we found that with our memory optimizations we could easily get to 64K context length.
> > What use case are you trying to work on? We could revisit our prioritization.
> 
> In my opinion, everyone is chasing after O1. In this process, the training on long texts in SFT (Supervised Fine-Tuning) and RL (Reinforcement Learning) is necessary.

I really can’t agree more！

**Comment by felipemello1** [comment]
thanks for the feedback! It seems that multiple users want it. We have plans to support tensor parallelism, which should help with that. But once we incorporate more types with parallelism with device mesh, it shouldnt be too hard to expand it to context parallel. It is something that we may have by H1, but unfortunately, not in the next couple of weeks.

**Comment by dz1iang** [comment]
> thanks for the feedback! It seems that multiple users want it. We have plans to support tensor parallelism, which should help with that. But once we incorporate more types with parallelism with device mesh, it shouldnt be too hard to expand it to context parallel.
> 
> TLDR: It is a feature that we may have by H1, but unfortunately, not in the next couple of weeks.

Hello. May I ask if there has been any recent progress regarding the CP ? We would like to use this feature. @felipemello1

**Comment by ebsmothers** [comment]
Hi @dz1iang are you interested in using context parallel as a standalone feature? Currently we have been holding off as the API doesn't yet compose with flex attention, which we rely on quite a bit (e.g. for dataset packing, Llama4 iRoPE, etc). However, if you don't care about using flex attention we could consider enabling and gating for the time being.

**Comment by dz1iang** [comment]
> Hi [@dz1iang](https://github.com/dz1iang) are you interested in using context parallel as a standalone feature? Currently we have been holding off as the API doesn't yet compose with flex attention, which we rely on quite a bit (e.g. for dataset packing, Llama4 iRoPE, etc). However, if you don't care about using flex attention we could consider enabling and gating for the time being.

@ebsmothers OK, thank you very much. I can try to use it without the support of flex attention. Additionally, I still hope that this feature can be supported as soon as possible, as long texts are quite common in post-training.

**Comment by ebsmothers** [comment]
Thanks @dz1iang, we are actively working on supporting this. Hope to get back to you with an update soon!

**Comment by dz1iang** [comment]
> Thanks [@dz1iang](https://github.com/dz1iang), we are actively working on supporting this. Hope to get back to you with an update soon!

Hi @ebsmothers, I would like to confirm if the CP is implemented based on this method (https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/examples/flex_attention_cp.py#L104). Thank you.

**Comment by Jantory** [comment]
Hi @felipemello1 , may I know what is the current progress on supporting long context SFT? And if you can provide some nice references about how to do SFT with data points that surpass the context window of the base model. I searched around but found little information.

**Comment by felipemello1** [comment]
hey @Jantory , i am not sure if this is 100% supported/working. You can give it a try: https://github.com/pytorch/torchtune/pull/2668

But we are not adding more features to torchtune, as we are working on a replacement focused on scalability. We only fix critical bugs now. Please take a look: https://github.com/pytorch/torchtune/issues/2883

**Comment by Jantory** [comment]
Thanks for the promptly response @felipemello1 ! As I am trying to benchmarking thins, I will take a look from here, and hope it works for my setting.
